youtube_dl/extractor/medaltv.py

```diff
@@ -0,0 +1,138 @@
+            'upload_date': '20201020',
+            'uploader_id': 10619174,
+        }
+    }, {
+        'url': 'https://medal.tv/clips/36787208',
+        'md5': 'b6dc76b78195fff0b4f8bf4a33ec2148',
+        'info_dict': {
+            'id': '36787208',
+            'ext': 'mp4',
+            'title': 'u tk me i tk u bigger',
+            'description': 'Medal,https://medal.tv/desktop/',
+            'uploader': 'Mimicc',
+            'timestamp': 1605580939,
+            'upload_date': '20201117',
+            'uploader_id': 5156321,
+        }
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        hydration_data = self._search_regex(
+            r'<script[^>]*>\s*(?:var\s*)?hydrationData\s*=\s*({.+?})\s*</script>',
+            webpage, 'hydration data', default='{}')
+        parsed = self._parse_json(hydration_data, video_id)
+
+        clip_info = try_get(parsed, lambda x: x['clips'][video_id], dict) or {}
+        if not clip_info:
+            raise ExtractorError('Could not find video information.',
+                                 video_id=video_id)
+
+        error = clip_info.get('error')
+        if error:
+            if(error == 404):
+                raise ExtractorError('That clip does not exist.',
+                                     expected=True, video_id=video_id)
+            else:
+                raise ExtractorError('An unknown error occurred ({0}).'.format(error),
+                                     video_id=video_id)
+
+        width = int_or_none(clip_info.get('sourceWidth'))
+        height = int_or_none(clip_info.get('sourceHeight'))
+
+        aspect_ratio = (width / height) if(width and height) else (16 / 9)
+
+        # ordered from lowest to highest resolution
+        heights = (144, 240, 360, 480, 720, 1080)
+
+        formats = []
+        thumbnails = []
+
+        for height in heights:
+            format_key = '{0}p'.format(height)
+            video_key = 'contentUrl{0}'.format(format_key)
+            thumbnail_key = 'thumbnail{0}'.format(format_key)
+            width = int(round(aspect_ratio * height))
+
+            # Second condition needed as sometimes medal says
+            # they have a format when in fact it is another format.
+            format_url = clip_info.get(video_key)
+            if(format_url and format_key in format_url):
+                formats.append({
+                    'url': format_url,
+                    'format_id': format_key,
+                    'width': width,
+                    'height': height
+                })
+
+            thumbnail_url = clip_info.get(thumbnail_key)
+            if(thumbnail_url and format_key in thumbnail_url):
+                thumbnails.append({
+                    'id': format_key,
+                    'url': thumbnail_url,
+                    'width': width,
+                    'height': height
+                })
```

Comment:
the process to extract the format and the thumbnail is similar, so these part needs to be abstracted to remove duplication.

Reply:
Okay, that is fair. Are we talking about adding something like:
```python
        info = {
            'formats': {
                'list': [],
                'key_name': 'contentUrl',
                'id_name': 'format_id'
            },
            'thumbnails': {
                'list': [],
                'key_name': 'thumbnail',
                'id_name': 'id'
            }
        }

        for height in heights:
            width = int(round(aspect_ratio * height))
            formatted_key = '{0}p'.format(height)

            for info_type in info:
                key = info[info_type]['key_name'] + formatted_key

                # Second condition needed as sometimes medal says
                # they have a format when in fact it is another format.
                url = clip_info.get(key)
                if(url and formatted_key in url):
                    info[info_type]['list'].append({
                        'url': url,
                        info[info_type]['id_name']: formatted_key,
                        'width': width,
                        'height': height
                    })
```

I originally chose to separate them, because youtube-dl requires different names for format ids, https://github.com/ytdl-org/youtube-dl/blob/7f41a598b3fba1bcab2817de64a08941200aa3c8/youtube_dl/extractor/common.py#L94-L303 as well as that the name templates were slightly different.

I could change the `info` dictionary to be a list, but the dictionary keys help name the items.

Reply:
kind of, i will try to abstract it further later(the `source` format also shares a bit code with this part).